semantic-reranker/src/main/java/org/opensearch/search/relevance/actionfilter/SearchActionFilter.java

```diff
@@ -0,0 +1,211 @@
+import org.opensearch.tasks.Task;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import static org.opensearch.action.search.ShardSearchFailure.readShardSearchFailure;
+
+public class SearchActionFilter implements ActionFilter {
+
+  // TODO: Finalize passage length and sliding window step
+  private static final int TOKENIZATION_LAMBDA_INPUT_SIZE_LIMIT = 600;
+  static final int SLIDING_WINDOW_STEP = TOKENIZATION_LAMBDA_INPUT_SIZE_LIMIT - 50;
+
+  // TODO: Update log levels where required
+  private static final Logger LOGGER = LogManager.getLogger(SearchActionFilter.class);
+
+  private final int order;
+
+  private final NamedWriteableRegistry namedWriteableRegistry;
+  private final SlidingWindowTextSplitter slidingWindowTextSplitter;
+
+  public SearchActionFilter() {
+    order = 10; // TODO: Finalize this value
+    namedWriteableRegistry = new NamedWriteableRegistry(Collections.emptyList());
+    slidingWindowTextSplitter = new SlidingWindowTextSplitter(TOKENIZATION_LAMBDA_INPUT_SIZE_LIMIT, SLIDING_WINDOW_STEP);
+  }
+
+  @Override
+  public int order() {
+    return order;
+  }
+
+  @Override
+  public <Request extends ActionRequest, Response extends ActionResponse> void apply(
+      final Task task,
+      final String action,
+      final Request request,
+      final ActionListener<Response> listener,
+      final ActionFilterChain<Request, Response> chain) {
+
+    // TODO: Double check tookTime calculation
+    final long startTime = System.nanoTime();
+
+    if (!SearchAction.INSTANCE.name().equals(action)) {
+      chain.proceed(task, action, request, listener);
+      return;
+    }
+
+    final SearchRequest searchRequest = (SearchRequest) request;
+    LOGGER.info("Applying action filter on search request: " + searchRequest);
+
+    final ActionListener<Response> searchResponseListener = getSearchResponseListener(searchRequest, listener, startTime);
+    chain.proceed(task, action, request, searchResponseListener == null? listener: searchResponseListener);
+  }
+
+  private <Response extends ActionResponse> ActionListener<Response> getSearchResponseListener(final SearchRequest request,
+      ActionListener<Response> listener, final long startTime) {
+
+    // TODO: Check index settings to get this value
+    final String semanticRankerEnabled = "true";
+    if (semanticRankerEnabled != null && !Boolean.valueOf(semanticRankerEnabled)) {
+      return null;
+    }
+
+    // Don't re-rank if scroll search is enabled
+    if (request.scroll() != null) {
+      return null;
+    }
+
+    // TODO: Should we re-rank if request source is empty?
+    if (request.source() == null) {
+      return null;
+    }
+
+    // TODO: Should we re-rank if it is a multi-index request?
+    final String[] indices = request.indices();
+    if (indices == null || indices.length != 1) {
```

Comment:
Minor: can we move these into a helper method to check if listener is applicable and call the method inside apply()?

Reply:
Yes, refactored the code